### ISSUE TITLE
onMessageListener not being called

### ISSUE THREAD

**Issue by hash1492** [QuickBlox/quickblox-javascript-sdk#57]
I am using quickblox Javascript SDK version 1.15.1 in my cordova app (using Ionic Framework)

There is a chat dialog between users A and B.
User A sends a message to user B, using the code shown below: 

```

$scope.sendMessage = function () {
            var data =  {
                send_to_chat:1,
                message: $scope.message.text,
                recipient_id: recipient_id
            }

            if(dialog_id){
                data.chat_dialog_id = dialog_id;
            }

            $http({method: 'POST', url: 'https://api.quickblox.com/chat/Message.json', headers: {'Content-Type': 'application/json','QB-Token': $rootScope.quickblox_session.token}, data: data})
            .then(function(response) {
                $scope.message.text = "";
                if(!dialog_id){
                    dialog_id = response.data.chat_dialog_id
                }
                getMessagesList();
            },function(err){
                console.log(err);
            })

        }

```

I have to use the REST API for sending message as the `QB.chat.send(opponentId, msg);` call throws an error  like "Cannot find send of undefined"

Anyways, the message is sent using the REST API call,  and it is added to the chat dialog. I think this should trigger the `onMessage` method that I have given to the onMessageListener

`QB.chat.onMessageListener = onMessage`

Basically whenever a user receives a message in a dialog, the listener should be called (Correct?). But it is not calling. But when I manually re-fetch the messages of this dialog, I do get the new message in that dialog

So the message is being added to the dialog but my listener is being called when i receive a new message. 

Please help me I am stuck on this for a long time! Thanks!

**Comment by soulfly** [comment]
You are right, but you have to add the 'send_to_chat' param

more info here http://quickblox.com/developers/Chat#Create_message

**Comment by hash1492** [comment]
Hey man thanks for replying! But as you can see above, I do have the "send_to_chat" param in my data object. So what am I doing wrong? The message is being added to the dialog, because when I manually re-fetch the messages of my dialog, I do get the new message also. 

But the listener is not being called. Please help. Thanks

**Comment by gremlinLee** [comment]
Good day. Could you provide additional info about how you init QB?

**Comment by hash1492** [comment]
```
var initializeQuickbox = function() {
    var CONFIG = {
        chatProtocol: {
            websocket: "wss://chat.quickblox.com:5291",
            active: 2
        },
        ssl: true,
        debug: {mode: 1}
    };


    // Initialize Quickblox
    QB.init(28732, "P-vdT2rhS-xUzCn", "BWpYvxTBaOWvbfN", CONFIG);
    // Get Facebook Access Token
    facebookConnectPlugin.getAccessToken(function(token) {
        var params = {provider: 'facebook', keys: {token: token}};
        // Create Quickblox session
        QB.createSession(params, function(err, result) {
            if (err) {
                console.log(err);
            }
            // I store the quickblox session object in localstorage over here so that I can use it in other pages
            ...
            ...
            ...

        });
    });
}

```

I am using Quickblox and connecting through Facebook access token rather than normal username password.

As soon as my app opens I initialize quickblox and create a session. I do this process in my base controller.
Then I fetch the list of dialogs, in another page (Works fine)
Then I click on a dialog so that it opens another page,and I can see its messages (Works fine)
When a user sends  a message to a dialog, I do send the "send_to_chat" param also in the post object. 

As soon as the message it added to the dialog, the onMessageListener should be called ideally. But it isn't being called.
But if I re-fetch the dialog's messages, the I do get the new message.

Is the procedure I followed correct?
1. Initilize Quickblox (please check my config for any mistake)
2. Create a session using Facebook token
3. Initialize the listener with `QB.chat.onMessageListener = onMessage` where onMessage is the method that should be called if I receive a message in that dialog;
4. Send the message with "send_to_chat" = 1

Thanks

**Comment by soulfly** [comment]
Use the following sequence to login to chat and send a message:

```
var user = {
      id: 4448514,
      name: 'chatuserweb1',
      login: 'chatuserweb1',
      pass: 'chatuserweb1'
    };

QB.createSession({login: user.login, password: user.pass}, function(err, res) {
  if (res) {
    QB.chat.connect({userId: user.id, password: user.pass}, function(err, roster) {
      if (err) {
          console.log(err);
      } else {

       /*
       *  (Object) roster - The user contact list
       *  roster = {
       *    '1126541': {subscription: 'both', ask: null},        // you and user with ID 1126541 subscribed to each other.
       *    '1126542': {subscription: 'none', ask: null},        // you don't have subscription but user maybe has
       *    '1126543': {subscription: 'none', ask: 'subscribe'}, // you haven't had subscription earlier but now you asked for it
       *  }; 
       */

      }
    });
  }else{
    console.log(err);
  }
});
```

```
var msg = {
  type: 'chat',
  body: "How are you today?",
  extension: {
    save_to_history: 1,
  }
};

var opponentId = 78;
QB.chat.send(opponentId, msg);

...

QB.chat.onMessageListener = onMessage;

function onMessage(userId, msg) {

}
```

**Comment by gremlinLee** [comment]
And wrong config

``` javascript
var CONFIG = {
        chatProtocol: {
            websocket: "wss://chat.quickblox.com:5291", // <- wrong, please replace with websockets
            active: 2
        },
        ssl: true,
        debug: {mode: 1}
    };
```

**Comment by hash1492** [comment]
@soulfly As I mentioned above, I am using Facebook access token in Quickblox. So the login and pass procedure you mentioned above is not for me I guess because the method `QB.chat.connect()`  requires an object with userId and password.

Also, this is the error I get when I use the `QB.chat.send(opponentId, msg);` method. That is why I have to use the REST API for sending the message

```
TypeError: Cannot read property 'send' of undefined
    at Object.a.Websocket._onIdle (quickblox.min.js:5)
    at Object.h.Connection._onIdle (quickblox.min.js:4)
    at Object.h.Connection.flush (quickblox.min.js:4)
    at Object.a.Websocket._send (quickblox.min.js:5)
    at Object.h.Connection.send (quickblox.min.js:4)
    at Object.d.send (quickblox.min.js:2)
    at l.$scope.sendMessage (individual-chat.ctrl.js:103)
    at ionic.bundle.min.js:229
    at ionic.bundle.min.js:386
    at l.$eval (ionic.bundle.min.js:156)

```

**Comment by hash1492** [comment]
@gremlinLee Hey I am a really new with this. I copied the config from the sample code so I am not sure what to do here. There is a property called "bosh", "websocket" and "active" and I am not sure which one I am supposed to use and what the value for each will be. I feel like I am really close. Could you tell me?

**Comment by soulfly** [comment]
You can login to chat also with Facebook user, just use session's token as a password as mentioned here

http://quickblox.com/developers/Chat#Password

```
QB.createSession(params, function(err, session) {
  if (res) {
    QB.chat.connect({userId: session.user_id, password: session.token}, function(err, roster) {
      if (err) {
          console.log(err);
      } else {

       /*
       *  (Object) roster - The user contact list
       *  roster = {
       *    '1126541': {subscription: 'both', ask: null},        // you and user with ID 1126541 subscribed to each other.
       *    '1126542': {subscription: 'none', ask: null},        // you don't have subscription but user maybe has
       *    '1126543': {subscription: 'none', ask: 'subscribe'}, // you haven't had subscription earlier but now you asked for it
       *  }; 
       */

      }
    });
  }else{
    console.log(err);
  }
});
```

**Comment by soulfly** [comment]
Can you also connect quickblox.js instead of quickblox.min.js

and post here a crash log when you call QB.chat.send(opponentId, msg);

**Comment by hash1492** [comment]
OMG thank you so much @soulfly !! Calling  `QB.chat.connect()` using the user_id and token as the password did the trick. Now the listener is being called. Thanks once again!
